src/components/modal/modal.test.tsx

```diff
@@ -0,0 +1,169 @@
+import React from "react";
+import { fireEvent, render, screen } from "@testing-library/react";
+
+import Modal from "./modal.component";
+import useScrollBlock from "../../hooks/__internal__/useScrollBlock";
+
+jest.mock("../../hooks/__internal__/useScrollBlock");
+const allowScroll = jest.fn();
+const blockScroll = jest.fn();
+
+const mockedUseScrollBlock = useScrollBlock as jest.MockedFunction<
+  typeof useScrollBlock
+>;
+
+mockedUseScrollBlock.mockReturnValue({
+  allowScroll,
+  blockScroll,
+});
+
+let onCancelFn: jest.Mock,
+  onCancelFnTwo: jest.Mock,
+  addEventListenerSpy: jest.SpyInstance,
+  removeEventListenerSpy: jest.SpyInstance;
+
+describe("event listeners", () => {
+  beforeEach(() => {
+    addEventListenerSpy = jest.spyOn(document, "addEventListener");
+    removeEventListenerSpy = jest.spyOn(document, "removeEventListener");
+  });
+
+  afterEach(() => {
+    addEventListenerSpy.mockRestore();
+    removeEventListenerSpy.mockRestore();
+  });
+
+  it("binds the key event listener to the document on mount", () => {
+    render(<Modal open onCancel={() => {}} />);
+
+    expect(addEventListenerSpy).toHaveBeenCalledWith(
+      "keyup",
+      expect.any(Function)
+    );
+  });
+
+  it("does not bind if component is not open on mount", () => {
+    render(<Modal open={false} onCancel={() => {}} />);
+
+    expect(addEventListenerSpy).not.toHaveBeenCalled();
+  });
+
+  it("removes the event listener if modal was open on unmount", () => {
+    const { unmount } = render(<Modal open onCancel={() => {}} />);
+
+    unmount();
+
+    expect(removeEventListenerSpy).toHaveBeenCalledTimes(1);
+  });
+
+  it("does not remove the event listener if it was not in use on unmount", () => {
+    const { unmount } = render(<Modal open={false} onCancel={() => {}} />);
+
+    unmount();
+
+    expect(removeEventListenerSpy).not.toHaveBeenCalled();
+  });
+
+  it("adds event listeners on modal open", () => {
+    const { rerender } = render(<Modal open={false} onCancel={() => {}} />);
+
+    expect(addEventListenerSpy).not.toHaveBeenCalled();
+
+    rerender(<Modal open onCancel={() => {}} />);
+
+    expect(addEventListenerSpy).toHaveBeenCalledTimes(1);
+  });
+
+  it("removes event listeners on modal close", () => {
+    const { rerender } = render(<Modal open onCancel={() => {}} />);
+
+    rerender(<Modal open={false} onCancel={() => {}} />);
+
+    expect(removeEventListenerSpy).toHaveBeenCalledTimes(1);
+  });
+});
+
+test("closes top modal when the `escape` key is pressed", () => {
+  onCancelFn = jest.fn();
+  onCancelFnTwo = jest.fn();
+
+  render(
+    <>
+      <Modal data-role="first-modal" open onCancel={onCancelFn} />
+      <Modal data-role="second-modal" open onCancel={onCancelFnTwo} />
+    </>
+  );
+
+  fireEvent.keyUp(document, { key: "Escape", code: "Escape" });
```

Comment:
**suggestion**: what do you think of using the user-event library here instead? The library would handle events in a manner that more closely resembles how events are handled in the browser. 

```suggestion
  await user.keyboard("{Escape}");
```

You may need to use Jest's fake timers if you agree, since `Modal` uses a third-party code (React Transition Group) which could schedule asynchronous tasks that are outside our control:

```ts
test("closes top modal when Escape key is pressed", async () => {
   jest.useFakeTimers();
   
   const firstOnCancel = jest.fn();
   const secondOnCancel = jest.fn();
   const user = userEvent.setup({ advanceTimers: jest.advanceByTime });

   // test code...

   jest.runOnlyPendingTimers(); // flush any potential scheduled tasks
   jest.useRealTimers();
});
```
For more detail on this, you can refer to[ RTL's advice on Fake Timers](https://testing-library.com/docs/using-fake-timers/).

Reply:
I did try to do it this way originally but I didn't do the `advanceTimers` so the tests either didn't pass or weren't reliable, hence the use of `fireEvent`. I've tried it your way and this works so I'll go with this.